Non fungible token (#131)
scale-it/algorand-builder: examples/nft/README.md

# Non-Fungible-Token Example using StateFul Teal

> In this example, we create a new non-fungible-token represented by a name and a url.

Reviewer: ```suggestion
In this example, we create a new non-fungible-token represented by a name and a url.
Compared to standard ASA, in this example we create a smart-contract to manage and store NFT balances. 

```

`algob` provides two functions :-
* `createNewNFT` - creates a new Non Fungible Token. Only the smart contract admin can create a new NFT.
* `transferNFT` - transfer an NFT from one account to another. Both account should have opted-in to the smart contract and the account wishing to transfer the NFT must hold the NFT (in the account's local state).

Also to be noted
* An account can hold upto a maximum of 16 NFT's in it's local storage.

> * The system can hold a maximum of 63 NFT's (in global storage)

Reviewer: Please add explanation to this limitations.

Author: ok

> * The system can hold a maximum of 63 NFT's (in global storage)

Reviewer: ```suggestion
* The NFT smart contract can hold a maximum of 63 NFT's (in global storage)
```

> # Non-Fungible-Token Example using StateFul Teal

Reviewer: ```suggestion
# Non-Fungible-Token Example using stateful TEAL
```

In this example, we create a new non-fungible-token represented by a name and a ref.
Compared to standard ASA, in this example we create a smart-contract to manage and store NFT balances. 

> - each NFT is an `(ID, ref_data, hash)` triple, Stored in the smart contract with the mapping of `{id: ref_data, id_h: hash}`
> - `id` is the NFT ID which is calculated as - (total count of nft in asc + 1). Eg. `nft_id = 1` for the first nft in the smart contract.
> - `ref_data` is represented by encoding nft details passed to smart contract. Eg. `name: "nft-1"`, `ref: some-ref` is passed to asc, then `ref_data` = `https://nft-1.com/<nft_id>/some-ref`.
> - `id_h` is a `Sha256` hash of  `ref`.

Reviewer: ```suggestion
- Each NFT is an `(ID, ref_data, hash)` triple, stored in the smart contract as `{id: ref_data, id_h: hash}` mapping.
- `id` is the NFT ID which is calculated as: `total_count_of_nft + 1`. Eg: `id = 1` for the first nft in the smart contract.
- `ref_data` is associated reference data. Eg: `ref_data` = `https://nft.com/<nft_id>/ref`.
- `id_h` is a `sha256` hash of  the content of reference data.
```

> Please check smart contract for the available commands and passing arguments to the smart contract.

Reviewer: ```suggestion
Please check smart contract for the available commands and arguments to the smart contract.
```

> Also to be noted

Reviewer: ```suggestion
NOTE:
```

* An account can hold upto a maximum of 8 NFT's in it's local storage.
* The NFT smart contract can hold a maximum of 32 NFT's (in global storage)
* _**This is only a proof of concept**. In production use we need to handle properly the account which can manage the smart-contract (update or delete)_. 
* In this ASC, Update Application call is **blocked** i.e once deployed, you cannot update the smart contract by another program. Refer to [this](https://developer.algorand.org/docs/features/asc1/stateful/#update-stateful-smart-contract) for more details. 

> This is because the ASC can store a maximum of 16 values (`int/[]byte`) in local storage and 64 values (`int/[]byte`)  in the global storage. Read more about state storage [here](https://developer.algorand.org/docs/features/asc1/stateful/sdks/#state-storage).

Reviewer: ```suggestion
This is because the ASC can store a maximum of 16 values (`int/[]byte`) in user account local storage and 64 values (`int/[]byte`)  in the global storage. Read more about state storage [here](https://developer.algorand.org/docs/features/asc1/stateful/sdks/#state-storage).
```
